packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart

```diff
@@ -139,32 +142,110 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
     Future<void> startSyncing() async {
       final batchStream = _getLogBatchesToSync();
       await for (final (logs, events) in batchStream) {
-        final response = await _sendToCloudWatch(events);
-        // TODO(nikahsn): handle tooOldLogEventEndIndex
-        // and expiredLogEventEndIndex.
-        if (response.rejectedLogEventsInfo?.tooNewLogEventStartIndex != null) {
-          // TODO(nikahsn): throw and exception to enable log rotation if the
-          // log store is full.
+        _TooNewLogEventException? tooNewException;
+        while (logs.isNotEmpty && events.isNotEmpty) {
+          final rejectedLogEventsInfo =
+              (await _sendToCloudWatch(events)).rejectedLogEventsInfo;
+          final expiredLogEventEndIndex =
+              rejectedLogEventsInfo?.expiredLogEventEndIndex;
+          final tooOldLogEventEndIndex =
+              rejectedLogEventsInfo?.tooOldLogEventEndIndex;
+          final tooNewLogEventStartIndex =
+              rejectedLogEventsInfo?.tooNewLogEventStartIndex;
+
+          if (expiredLogEventEndIndex != null &&
+              expiredLogEventEndIndex >= 0 &&
+              expiredLogEventEndIndex <= events.length - 1) {
+            // delete old logs from log store
+            await _logStore
+                .deleteItems(logs.sublist(0, expiredLogEventEndIndex + 1));
+            // set logs to start from next
+            logs.removeRange(0, expiredLogEventEndIndex + 1);
+            // set events to start from next
+            events.removeRange(0, expiredLogEventEndIndex + 1);
+            continue;
+          }
+          if (tooOldLogEventEndIndex != null &&
+              tooOldLogEventEndIndex >= 0 &&
+              tooOldLogEventEndIndex <= events.length - 1) {
+            // delete old logs from log store
+            await _logStore
+                .deleteItems(logs.sublist(0, tooOldLogEventEndIndex + 1));
+            // set logs to start from next
+            logs.removeRange(0, tooOldLogEventEndIndex + 1);
+            // set events to start from next
+            events.removeRange(0, tooOldLogEventEndIndex + 1);
+            continue;
+          }
+          if (tooNewLogEventStartIndex != null &&
+              tooNewLogEventStartIndex >= 0 &&
+              tooNewLogEventStartIndex <= events.length - 1) {
+            tooNewException = _TooNewLogEventException(
```

Comment:
Why not throw it here? The control flow would be easier to follow

Reply:
we want to continue with syncing the current batch after sanitizing the events in the current batch and then throw to stop syncing next batches